Migrating from process.binding('config') to getOptions()

burgerboydaddy opened the pull request:
Change inside /lib/internal/bootstrap/node.js code from
process.binding('config') to
internalBinding('options).getOptions()

Changes are done for options:
--experimental-modules
--experimental-vm-modules
--experimental-worker

##### Checklist

- [x ] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes

devsnek commented on lib/internal/bootstrap/node.js:
```diff
@@ -137,7 +137,7 @@
     }
 
-    if (process.binding('config').experimentalWorker) {
+    if (internalBinding('options').getOptions('--experimental-worker')) {
```
can you factor out the `internalBinding` into a single call up top and also save the flag calls as variables to be used later?

something like 
```js
const { getOptions } = internalBinding('options');
const experimentalModules = getOptions('experimentalModules');
// ...
if (experimentalModules) {
  // ...
```

burgerboydaddy replied:
Yes I will do that.
But one question. There is already some code in place like line 114:
`const options = internalBinding('options');
    if (options.getOptions('--help')) {
      NativeModule.require('internal/print_help').print(process.stdout);
      return;
    }`

Should I redo that part also and change code to use as mentioned:
`const { getOptions } = internalBinding('options');
const experimentalModules = getOptions('experimentalModules');
// ...
if (experimentalModules) {
  // ...`

Thanks

burgerboydaddy replied:
As mentioned in my previous comment, will code like this (replacing line 114) be Ok:
```
    const { getOptions } = internalBinding('options');
    const helpOption = getOptions('--help');
    const completionBashOption = getOptions('--completion-bash');
    const experimentalModulesOption = getOptions('--experimental-modules');
    const experimentalVMModulesOption = getOptions('--experimental-vm-modules');
    const experimentalWorkerOption = getOptions('--experimental-worker');
```

addaleax replied:
@burgerboydaddy Yes, that definitely looks okay :)

burgerboydaddy commented:
Updated code as per suggestions
```
diff --git a/lib/internal/bootstrap/node.js b/lib/internal/bootstrap/node.js
index f53ed55dcad..7681c81bd4b 100644
--- a/lib/internal/bootstrap/node.js
+++ b/lib/internal/bootstrap/node.js
@@ -111,13 +111,18 @@
       NativeModule.require('internal/inspector_async_hook').setup();
     }
 
-    const options = internalBinding('options');
-    if (options.getOptions('--help')) {
+    const { getOptions } = internalBinding('options');
+    const helpOption = getOptions('--help');
+    const completionBashOption = getOptions('--completion-bash');
+    const experimentalModulesOption = getOptions('--experimental-modules');
+    const experimentalVMModulesOption = getOptions('--experimental-vm-modules');
+    const experimentalWorkerOption = getOptions('--experimental-worker');
+     if (helpOption) {
       NativeModule.require('internal/print_help').print(process.stdout);
       return;
     }
 
-    if (options.getOptions('--completion-bash')) {
+    if (completionBashOption) {
       NativeModule.require('internal/bash_completion').print(process.stdout);
       return;
     }
@@ -137,7 +142,7 @@
       setupQueueMicrotask();
     }
 
-    if (process.binding('config').experimentalWorker) {
+    if (experimentalWorkerOption) {
       setupDOMException();
     }
 
@@ -169,9 +174,9 @@
         'DeprecationWarning', 'DEP0062', startup, true);
     }
 
-    if (process.binding('config').experimentalModules ||
-        process.binding('config').experimentalVMModules) {
-      if (process.binding('config').experimentalModules) {
+    if (experimentalModulesOption ||
+      experimentalVMModulesOption) {
+      if (experimentalModulesOption) {
         process.emitWarning(
           'The ESM module loader is experimental.',
           'ExperimentalWarning', undefined);
@@ -183,7 +188,7 @@
     {
       // Install legacy getters on the `util` binding for typechecking.
       // TODO(addaleax): Turn into a full runtime deprecation.
-      const { pendingDeprecation } = process.binding('config');
+      const { pendingDeprecation } = internalBinding('options').getOptions();
       const utilBinding = internalBinding('util');
       const types = internalBinding('types');
       for (const name of [
@@ -868,4 +873,4 @@
   }
 
   startup();
-});
+});
\ No newline at end of file
```

burgerboydaddy commented:
Added suggested changes to the code
```
const { getOptions } = internalBinding('options');
    const helpOption = getOptions('--help');
    const completionBashOption = getOptions('--completion-bash');
    const experimentalModulesOption = getOptions('--experimental-modules');
    const experimentalVMModulesOption = getOptions('--experimental-vm-modules');
    const experimentalWorkerOption = getOptions('--experimental-worker');
```

addaleax commented on lib/internal/bootstrap/node.js:
```diff
@@ -183,7 +188,7 @@
       // Install legacy getters on the `util` binding for typechecking.
       // TODO(addaleax): Turn into a full runtime deprecation.
-      const { pendingDeprecation } = process.binding('config');
+      const { pendingDeprecation } = internalBinding('options').getOptions();
```
@burgerboydaddy I think something went wrong with this one?

burgerboydaddy replied:
@addaleax initial request was to replace process.binding('config') to
internalBinding('options).getOptions(). Is there any reason why this line shouldn't be included?
If yes, I will restore this line to original one.
Please let me know what is your opinion.

thanks

burgerboydaddy commented:
Fixed issues.
returned `process.binding('config');` in line 191:
-      const { pendingDeprecation } = internalBinding('options').getOptions();
+      const { pendingDeprecation } = process.binding('config');

addaleax commented:
@burgerboydaddy Something seems to be odd here … github says there are about 200 commits in this PR, some of which are merge commits?

Do you think you could rebase this against `master` (not `merge`)?

burgerboydaddy commented:
@addaleax I'm even thinking to re-do complete request (new PR). Will that be better?

addaleax commented:
@burgerboydaddy You can do that, but rebasing this one should be totally fine as well – whichever you prefer

burgerboydaddy commented:
> @burgerboydaddy You can do that, but rebasing this one should be totally fine as well – whichever you prefer

@addaleax This are command that I just executed:
```
$ git fetch --all
$ git rebase origin/master
```

During rebase got error:
```
Using index info to reconstruct a base tree...
M	lib/internal/bootstrap/node.js
Falling back to patching base and 3-way merge...
No changes -- Patch already applied.
Applying: Updated code for Migrating from process.binding('config') to getOptions() #23588
Applying: Migrating from process.binding('config') to getOptions
Using index info to reconstruct a base tree...
M	lib/internal/bootstrap/node.js
Falling back to patching base and 3-way merge...
Auto-merging lib/internal/bootstrap/node.js
CONFLICT (content): Merge conflict in lib/internal/bootstrap/node.js
error: Failed to merge in the changes.
Patch failed at 0193 Migrating from process.binding('config') to getOptions
Use 'git am --show-current-patch' to see the failed patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".

```

I fixed issue with file that I changed (node.js). After that did:

```
git add lib/internal/bootstrap/node.js
git rebase --continue
```

But after rebase --continue received response:

```
Applying: Migrating from process.binding('config') to getOptions
No changes - did you forget to use 'git add'?
If there is nothing left to stage, chances are that something else
already introduced the same changes; you might want to skip this patch.
```

What should I do?

addaleax commented:
@burgerboydaddy That’s odd … it would typically mean that the changes from your commits have already happened upstream; but that’s clearly not the case, so I’m not sure what to do with that.

burgerboydaddy commented:
@addaleax ok, how about to close this PR and create new one with proper code? I will mention this id (#23588) in my comments?

Trott commented:
Rebased, addressed merge conflict, force pushed.

Trott commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/18054/

Trott commented:
Resume Build: https://ci.nodejs.org/job/node-test-pull-request/18063/

Trott commented:
Resume Build: https://ci.nodejs.org/job/node-test-pull-request/18080/

Trott commented:
Just landed a PR to fix the unreliable test that failed last CI run...

Resume Build CI: https://ci.nodejs.org/job/node-test-pull-request/18094/

Trott commented:
Landed in bb79e76.

Thanks for the contribution! 🎉

(If you're interested in other possible contributions to Node.js but don't have a good idea of where to start looking, some ideas are posted at https://www.nodetodo.org/next-steps/.)
